## Problem statement 



1.Amidst raising of obesity rate and sedatory lifestyle,irrespective of age people face the type1 and type2 diabetes.Since the increased amount of patients willing to ensure that they contain diabetes or vice versa. The manual process of detecting diabetes from the reports requires more time and effort. So that the problem statement received from the business side is to develop the ANN model inorder to predict either the patient suffering from diabetes or not.

2.Possible outcome of target feature for this problem statement and further observation resulted that this problem statement can be solved by using Binary classification. 

In [3]:
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from sklearn.preprocessing import StandardScaler

#     ===============================================

##  Data Acquistion

### downloading the data directly

In [4]:
diabetes_data = pd.read_csv(r'C:\Users\Gobinath\Desktop\tasks\diabetes infference\diabetes.csv')
diabetes_data

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

#### Observations:
1.The size of the data is 768 observations and 9 features

2.Details of the features:

Pregnancies: Number of times pregnant

Glucose: Plasma glucose concentration a 2 hours in an oral glucose tolerance test

BloodPressure: Diastolic blood pressure (mm Hg)

SkinThickness: Triceps skin fold thickness (mm)

Insulin: 2-Hour serum insulin (mu U/ml)

BMI: Body mass index (weight in kg/(height in m)^2)

DiabetesPedigreeFunction: Diabetes pedigree function

Age: Age (years)

Outcome: Class variable (0 or 1)


In [5]:
thisdict ={
  "Pregnancies": "Number of times pregnant",
  "Glucose": "Plasma glucose concentration a 2 hours in an oral glucose tolerance test",
  "BloodPressure:": "Diastolic blood pressure (mm Hg)"
  "SkinThickness": "Triceps skin fold thickness (mm)"
  "Insulin": "2-Hour serum insulin (mu U/ml)"
  "BMI": "Body mass index (weight in kg/(height in m)^2)"
  "DiabetesPedigreeFunction": "Diabetes pedigree function"
  "Age": "Age (years)"
  "Outcome": "Class variable (0 or 1)"
}

SyntaxError: invalid character in identifier (<ipython-input-5-018657b1181d>, line 2)

## Data Understanding

In [ ]:
diabetes_data.isna().sum()

In [ ]:
diabetes_data.dtypes

### Identify the target variables.
1.Outcome is observed to be the output feature.

2.Seperating the output feature is performed in the below cell.

In [ ]:
X = diabetes_data.drop('Outcome',axis = 1)
y = diabetes_data[['Outcome']]

In [ ]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
mm_scaler = MinMaxScaler()
scaled_X = mm_scaler.fit_transform(X)
scaled_X = pd.DataFrame(scaled_X,columns=X.columns)

### Split the data into training set and testing set

In [ ]:
X_train,X_test,y_train,y_test = train_test_split(scaled_X,y,test_size=0.20,stratify=y,random_state=123)

In [ ]:
X_train.shape,y_train.shape #Training Samples

In [ ]:
X_test.shape,y_test.shape #Test samples

### Report
Mention the method adopted and justify why the method was used

to remove duplicate data, if present -- None

to impute or remove missing data, if present-- None

to remove data inconsistencies, if present -- None

to encode categorical data -- None

the normalization technique used -- MinMaxScaler

## Design the architecture

In [ ]:
from keras.models import Sequential
from keras.layers import Dense
import keras

In [ ]:
model = Sequential()
model.add(Dense(units = 20,input_dim = 8,activation='relu',kernel_initializer = 'glorot_uniform'))
model.add(Dense(units = 6,activation='relu',kernel_initializer = 'glorot_uniform'))
model.add(Dense(units = 1,activation='sigmoid',kernel_initializer = 'glorot_uniform'))
model.summary()

### observation: 

1.The model used will be sequential (i.e, connected to one another in ordered manner),

2.dense layers is used to construct the model,

3.ReLu is the adaptive activation method used to handle the features.

4.Since it is the binary classification one neuron is used as the output to predict whether patient has diabetes or not.


## compiling the model

In [ ]:
gradient_descent_opt = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9, nesterov=True)
adagrad = tf.keras.optimizers.Adagrad(learning_rate=0.01,initial_accumulator_value=0.1)

In [ ]:
model.compile(optimizer=adagrad,loss="binary_crossentropy",metrics=['accuracy'])

## Train the architecture

In [ ]:
model_training = model.fit(x=X_train,y=y_train, batch_size=32,epochs=10,verbose=2)

### Finding the optimal hyperparameters

In [ ]:
from sklearn.model_selection import GridSearchCV, KFold
from keras.wrappers.scikit_learn import KerasClassifier

In [ ]:
def create_model():
    model = Sequential()
    model.add(Dense(12, input_dim=8, kernel_initializer='uniform', activation='relu'))
    model.add(Dense(8,kernel_initializer='uniform', activation='relu'))
    model.add(Dense(1, kernel_initializer='uniform', activation='sigmoid'))
    #adam=adam_v2(lr=0.01)
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

In [ ]:
model = KerasClassifier(build_fn = create_model,verbose = 0)

In [ ]:
grid = GridSearchCV(estimator = model,param_grid = {'batch_size' : [10,20,40],'epochs' : [10,20,30]},
                    cv = KFold(),
                    verbose = 10)
grid_result = grid.fit(X_train,y_train)

### observation:
1. Above Gridsearch function result us the suitable number of epochs and batch size to be used.

In [ ]:
print('Best : {}, using {}'.format(grid_result.best_score_,grid_result.best_params_))

In [ ]:
means = grid_result.cv_results_['mean_test_score']
stds  = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']

In [ ]:
grid_result.cv_results_['mean_test_score']

In [ ]:
grid_result.cv_results_

In [ ]:
means

In [ ]:
for mean, stdev, param in zip(means, stds, params):
    print('{},{} with: {}'.format(mean, stdev, param))

### observation:
1.Above observation results us the consolidated number of batches and number of epochs to be used. 

In [ ]:
import warnings
warnings.filterwarnings('ignore')

In [ ]:
# Defining the model

def create_model(neuron1,neuron2):
    model = Sequential()
    model.add(Dense(neuron1,input_dim = 8,kernel_initializer = 'uniform',activation = 'linear'))
    model.add(Dropout(0.2))
    model.add(Dense(neuron2,input_dim = neuron1,kernel_initializer = 'uniform',activation = 'linear'))
    model.add(Dropout(0.2))
    model.add(Dense(1,activation = 'sigmoid'))
    
    #adam = Adam(lr = 0.001)
    model.compile(loss = 'binary_crossentropy',optimizer = 'adam',metrics = ['accuracy'])
    return model

# Create the model

model = KerasClassifier(build_fn = create_model,verbose = 0,batch_size = 40,epochs = 50)

# Define the grid search parameters

neuron1 = [4,8,16]
neuron2 = [2,4,8]

# Make a dictionary of the grid search parameters

param_grids = dict(neuron1 = neuron1,neuron2 = neuron2)

# Build and fit the GridSearchCV

grid = GridSearchCV(estimator = model,param_grid = param_grids,cv = KFold(),verbose = 10)
grid_result = grid.fit(X_train,y_train)

### observation:
1. Above automated method gives us the suitable number of neurons to be used in the each layer of the architecture 

In [ ]:
print('Best : {}, using {}'.format(grid_result.best_score_,grid_result.best_params_))

means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']

for mean, stdev, param in zip(means, stds, params):
      print('{},{} with: {}'.format(mean, stdev, param))

### observation:
1.Filtering out the optimal number of neurons that can be used.

In [ ]:
from sklearn.metrics import classification_report, accuracy_score

# Defining the model

def create_model():
    model = Sequential()
    model.add(Dense(16,input_dim = 8,kernel_initializer = 'uniform',activation = 'linear'))
    model.add(Dropout(0.1))
    model.add(Dense(8,kernel_initializer = 'uniform',activation = 'linear'))
    model.add(Dropout(0.1))
    model.add(Dense(1,activation = 'sigmoid'))
    
    #adam = Adam(lr = 0.001) #sgd = SGD(lr=learning_rate, momentum=momentum, decay=decay_rate, nesterov=False)
    model.compile(loss = 'binary_crossentropy',optimizer = 'adam',metrics = ['accuracy'])
    return model

# Create the model

model = KerasClassifier(build_fn = create_model,verbose = 0,batch_size = 40,epochs = 10)

# Fitting the model

model.fit(X_train,y_train)

# Predicting using trained model

y_predict = model.predict(X_test)

# Printing the metrics
print(accuracy_score(y_test,y_predict))

In [6]:
print(classification_report(y_test,y_predict))

ERROR:root:Internal Python error in the inspect module.
Below is the traceback from this internal error.



Traceback (most recent call last):
  File "C:\Users\Gobinath\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-6-bc5f41cc2a3c>", line 1, in <module>
    print(classification_report(y_test,y_predict))
NameError: name 'classification_report' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Gobinath\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py", line 2040, in showtraceback
    stb = value._render_traceback_()
AttributeError: 'NameError' object has no attribute '_render_traceback_'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Gobinath\Anaconda3\lib\site-packages\IPython\core\ultratb.py", line 1101, in get_records
    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)
  File "C:\U

NameError: name 'classification_report' is not defined

### Conclusion

#### Above classification used for the measurement for the accuracy of the predicitions of the  model.
1. The proposed has clarity about the classification of the non-diabetic patients.

2.About the diabetic classification our model, is relatively low.

3. So further hyperparameter tuning is required, to find out suitable combination of hyperparameters to classify diabetic patients.

### Solutions


1.Inorder to solve this business problem, apart from neural nets, efficient ML models can be used, because those light weight models able to classify the results with limited number of features.

2.Secondly the more features and observations can be requested from the business side, inorder to provide enough details to the neural nets to make classification efficiently.

3.my learning involves the automation of the hyper parameter tuning and understanding of the efficiency of the normalization concept.

4.observations include usage of the different activation functions at suitable scenarios.

5.declaring the suitable number of neurons in the input and hidden and output layers through automation.

6.decision made include the selection of the lower number of epochs and batch size with proven record of increased accuracy through the automation of selection of suitable hyperparameters.



